pkg/api/v1/atlasfederatedauth_types.go

```diff
@@ -0,0 +1,155 @@
+package v1
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.mongodb.org/atlas/mongodbatlas"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1/common"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1/status"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/util/kube"
+)
+
+func init() {
+	SchemeBuilder.Register(&AtlasFederatedAuth{}, &AtlasFederatedAuthList{})
+}
+
+type AtlasFederatedAuthSpec struct {
+	// +kubebuilder:default:=false
+	Enabled *bool `json:"enabled,omitempty"`
+	// Connection secret with API credentials for configuring the federation.
+	// These credentials must have OrganizationOwner permissions.
+	ConnectionSecretRef common.ResourceRefNamespaced `json:"connectionSecretRef,omitempty"`
+	// Approved domains that restrict users who can join the organization based on their email address.
+	// +optional
+	DomainAllowList []string `json:"domainAllowList,omitempty"`
+	// Prevent users in the federation from accessing organizations outside of the federation, and creating new organizations.
+	// This option applies to the entire federation.
+	// See more information at https://www.mongodb.com/docs/atlas/security/federation-advanced-options/#restrict-user-membership-to-the-federation
+	// +kubebuilder:default:=false
+	DomainRestrictionEnabled *bool `json:"domainRestrictionEnabled,omitempty"`
+	// +kubebuilder:default:=false
+	// +optional
+	SSODebugEnabled *bool `json:"ssoDebugEnabled,omitempty"`
+	// Atlas roles that are granted to a user in this organization after authenticating.
+	// +kubebuilder:validation:Enum=ORG_MEMBER;ORG_READ_ONLY;ORG_BILLING_ADMIN;ORG_GROUP_CREATOR;ORG_OWNER;ORG_BILLING_READ_ONLY;ORG_TEAM_MEMBERS_ADMIN
+	// +optional
+	PostAuthRoleGrants []string `json:"postAuthRoleGrants,omitempty"`
+	// Map IDP groups to Atlas roles.
+	// +optional
+	RoleMappings []RoleMapping `json:"roleMappings,omitempty"`
+}
+
+func (f *AtlasFederatedAuthSpec) ToAtlas(orgID, idpID string, client *mongodbatlas.Client) (*mongodbatlas.FederatedSettingsConnectedOrganization, error) {
```

Comment:
Usually `ToAtlas` methods are plain local conversions, but here we are passing a client, which means this conversion depends on an external system. I would rather make it a simple self contained conversion by passing the the pre-fetched projects by role:

```suggestion
func (f *AtlasFederatedAuthSpec) ToAtlas(orgID, idpID string, projectsByRole map[string]*mongobdatlas.Project) (*mongodbatlas.FederatedSettingsConnectedOrganization, error) {
```

You would have a separate method to produce the map of projects using the client.